2_idioms/src/main.rs

```diff
@@ -1,3 +1,452 @@
+            println!(
+                "product:{}, price:{}, in stock:{:?};",
+                product[0].name(),
+                product[0].value(),
+                product.len()
+            )
+        })
+    }
+
+    pub fn insert<T: Product>(&mut self, items: Vec<Box<T>>) -> Result<(), &'static str> {
+        let tid = TypeId::of::<T>();
+        match self.product_storage.get_mut(&tid) {
+            Some(product) => {
+                if product.len() + items.len() > self.in_stock_capacity.0 as usize {
+                    return Err("Cannot add that much");
+                }
+                for item in items {
+                    product.push(item)
+                }
+                Ok(())
+            }
+            None => {
+                if self.product_storage.len() == self.stock_capacity.0 as usize {
+                    return Err("There's no empty positions in stock");
+                }
+                if items.len() > self.in_stock_capacity.0 as usize {
+                    return Err("Cannot add that much");
+                }
+                let mut new_vec: Vec<Box<dyn Product>> = Vec::new();
+                for item in items {
+                    new_vec.push(item);
+                }
+                self.product_storage.insert(TypeId::of::<T>(), new_vec);
+                return Ok(());
+            }
+        }
+    }
+
+    pub fn get_price_and_count<T: Product>(&self) -> Option<(u64, usize)> {
+        let tid = TypeId::of::<T>();
+        match self.product_storage.get(&tid) {
+            None => None,
+            Some(product) => Some((product[0].value(), product.len())),
+        }
+    }
+
+    pub fn purchase<T: Product, TAmmount: Into<product::Amount>>(
+        &mut self,
+        amount: TAmmount,
+    ) -> Vec<Box<dyn Product>> {
+        let tid = TypeId::of::<T>();
+        let count = amount.into().0 as usize;
+        let items = self.product_storage.get_mut(&tid).unwrap();
+
+        let drain = items.drain(0..count);
+        let mut vec = Vec::new();
+        for item in drain {
+            vec.push(item);
+        }
+        if items.len() == 0 {
+            self.product_storage.remove(&tid);
+        }
+        vec
+    }
+}
+struct VendingMachine {
+    stock: Stock,
+    rest: Money,
+}
+#[derive(Clone, Copy)]
+struct RowCount(u8);
+#[derive(Clone, Copy)]
+struct InRowCount(u8);
+impl VendingMachine {
+    pub fn display_stock(&self) {
+        self.stock.display_stock()
+    }
+    pub fn new(stock_capacity: RowCount, in_stock_capacity: InRowCount) -> Self {
+        VendingMachine {
+            stock: Stock::new(stock_capacity, in_stock_capacity),
+            rest: Money::new(),
+        }
+    }
+
+    pub fn insert<T: Product, TCollection: IntoIterator<Item = T>>(
+        &mut self,
+        items: TCollection,
+    ) -> Result<(), &'static str> {
+        let mut collection = Vec::new();
+        for item in items {
+            collection.push(Box::new(item))
```

Comment:
@ArsileLuci use `.collect::<Vec<_>>()` instead of manual assemblying the vector.